simphony/engine/extension.py

```diff
@@ -0,0 +1,204 @@
+        """
+
+    @abstractmethod
+    def create_wrapper(self, cuds, engine_name, engine_interface):
+        """Creates a wrapper to the requested engine.
+
+        Parameters
+        ----------
+        cuds: CUDS
+          CUDS computational model data
+        engine_name: str
+          name of the engine, must be supported by this extension
+        engine_interface: EngineInterface
+          the interface to interact with engine
+
+        Returns
+        -------
+        ABCEngineExtension: A wrapper configured with cuds and ready to run
+        """
+
+    @staticmethod
+    def create_engine_metadata(name, features, interfaces):
+        """Factory method to create engine metadata.
+
+        Parameters
+        ----------
+        name: str
+          fixed name of a support engine to be used in user's data, i.e. key.
+        features: list
+          features of this engine as a list of EngineFeatureMetadata objects
+        interfaces: list
+          supported engine interfaces as a list of EngineInterface enums
+
+        Returns
+        -------
+        EngineMetadata: a loaded engine metadata object
+        """
+        return EngineMetadata(name, features, interfaces)
+
+    @staticmethod
+    def create_engine_metadata_feature(physics_equation, methods):
+        """Factory method to create engine metadata feature.
+
+        Parameters
+        ----------
+        physics_equation: PhysicsEquation
+          represents a physics equation
+        methods: list
+          ComputationalMethod objects to solve the equation
+
+        Return
+        ------
+        EngineFeatureMetadata: a loaded engine metadata feature object
+        """
+        return EngineFeatureMetadata(physics_equation, methods)
+
+
+class EngineManager(object):
+    """Controller class to keep track of engine extensions."""
+    def __init__(self):
+        self._engine_extensions = {}
+
+    def load_metadata(self, plugin):
+        """Load existing engine extensions into the manager.
+
+        This method inspects the given module and looks for ABCEngineExtension
+         subclasses. If provided, will instantiate and keep them for further
+         interactions with corresponding engines.
+
+        plugin: module
+          A python module provided by an extension
+        """
+        if not inspect.ismodule(plugin):
+            raise EngineManagerException(
+                'The provided object is not a module: %s' % plugin)
+
+        for name, value in inspect.getmembers(plugin, inspect.isclass):
+            if not inspect.isabstract(value) and\
+                    issubclass(value, ABCEngineExtension):
+                # Load the corresponding engine extension
+                extension = value()
+                if not isinstance(extension, ABCEngineExtension):
+                    raise ValueError('Expected ABCEngineExtension, got %s' %
+                                     extension)
```

Comment:
`isinstance` checks the result of invoking `value()` which might be of different type from `value` itself, for any reason. One example is overring __new__ method. I know, I am a little paranoid here!